Sources/ProcedureKit/ProcedureFuture.swift

```diff
@@ -0,0 +1,336 @@
+//
+//  ProcedureKit
+//
+//  Copyright © 2016 ProcedureKit. All rights reserved.
+//
+
+import Foundation
+import Dispatch
+
+// MARK: - ProcedurePromise & ProcedureFuture
+
+/**
+ A ProcedurePromise provides a ProcedureFuture that is signaled when the promise is completed.
+ 
+ Create a ProcedurePromise and retrieve the future from the promise:
+ 
+ ```swift
+ let promise = ProcedurePromise()
+ let future = promise.future // retrieve the future from the promise
+ ```
+ 
+ then after your asynchronous task is complete, complete the promise like so:
+ 
+ ```swift
+ promise.complete()
+ ```
+ 
+ and anything scheduled using the ProcedureFuture will occur after the promise has been completed.
+ 
+ A simple example is:
+ 
+ ```swift
+ func myAsyncFunction() -> ProcedureFuture {
+     let promise = ProcedurePromise()
+     // dispatch an asynchronous task
+     DispatchQueue.global().async {
+         // do something
+         // then complete the promise when done
+         promise.complete()
+     }
+     return promise.future
+ }
+ 
+ myAsyncFunction().then(on: DispatchQueue.global()) {
+    // execute this block after myAsyncFunction's promise completes, on DispatchQueue.global()
+ }
+ ```
+ 
+ A ProcedurePromise/ProcedureFuture does not return a value - it merely allows you to execute
+ a block when a promise is complete.
+ 
+ To return a future value, use ProcedureResultPromise (and ProcedureResultFuture).
+ */
+public class ProcedurePromise {
+    public let future = ProcedureFuture()
+    #if DEBUG
+    private var _didComplete = false
+    private var lock = PThreadMutex()
+    #endif
+
+    #if DEBUG
+    // in Debug builds, verify and assert if a promise is not completed before it is deinited
```

Comment:
Just to make sure you know, `assert`s are only compiled in when using `-O0`. I know you're also using `DEBUG` to add properties in such, just wanted to make sure you know.